Fixes #28810 - added validations for taxonomy's name

nofaralfasi opened the pull request:
Added validations to prevent from an organization to be named `Any Organization` and from a location to be named `Any Location`.  
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

adamruzicka commented:
While this is probably a step in the right direction, it solves only a part of the problem. The issue still exists if I use any non-english locale. For example if I switch my locale to czech, then "Any organization" becomes "Libovolná organizace", yet I am still allowed to create an organization with the name "Libovolná organizace" and then this patch doesn't really help.

As a counter proposal, how about instead of forbidding certain names, we somehow mark the "Any organization" context so it is distinctive and even if one has an organization with name "Any organization", one can be still able to tell them apart?

evgeni commented:
ok to test

ares commented:
> As a counter proposal, how about instead of forbidding certain names, we somehow mark the "Any organization" context so it is distinctive and even if one has an organization with name "Any organization", one can be still able to tell them apart?

Do you have any specific suggestion? cc @MariSvirik any thoughts on how to visually distinguish "Any organization" meaning no value selected from the organization named "Any organization" in the orgnization selector? Similarly for locations of course.

MariSvirik commented:
So when selecting org./ location it's evident that "Any organization/ location" is visibly different (There is a divider). Which is good. The problem is we don't see this division in the [context selector](https://www.patternfly.org/v4/components/context-selector/design-guidelines) when it's not expanded.

**Foreman implementation**
<img width="363" alt="Screenshot 2022-05-19 at 13 22 00" src="https://user-images.githubusercontent.com/51960547/169282964-7213867b-e947-443d-8fd0-0b76225bebca.png">

**PF4 Example**
<img width="165" alt="Screenshot 2022-05-19 at 13 34 39" src="https://user-images.githubusercontent.com/51960547/169284183-7107d872-4abe-412e-8860-3bf51fe479f4.png">

As you can see our implementation is a bit different from the Pf4 - a checkmark on the left side a not on the right. If we have a checkmark on the right we can add an icon - saying that is a real "any" setting. 
e.g. globe, building (anything from font awesome is possible)

Quick sketch here: 
<img width="476" alt="Screenshot 2022-05-19 at 14 15 57" src="https://user-images.githubusercontent.com/51960547/169291093-cebd5215-6e0b-49d4-9833-3f2ad93a96e8.png">

Of course, there wouldn't be a white context selector...
Any thoughts?

adamruzicka commented:
>  Any thoughts?

Yeah, I imagined something like that, although I was a lot less imaginative with the icons. Maybe just a funnel (and slashed funnel or an outline of it) marking filtered/unfiltered would work?

nofaralfasi commented:
I added a globe icon next to the `Any Organization` and `Any Location` texts, as you can see in the screenshot below:
![image](https://user-images.githubusercontent.com/38184193/170009752-f4c4d28a-d7af-4bf5-8afb-ea8f65d88def.png)

Unfortunately I am not able to add the icon to the texts in the top menu.

MariSvirik commented:
I would say it's crucial to have it for the selected item on top. Let's ask in a PF4 react slack. It shouldn't be that difficult, but I don't know what component you are using. 
For example there is an icon within dropdown [here](patternfly.org/v4/components/menu-toggle/#with-iconimage-and-text). (May not apply to our problem)

Also, would it be possible to fix capitalization? We should use [sentence case](https://www.patternfly.org/v4/ux-writing/capitalization#capitalization-across-red-hat-uis) - so "Any organization", "Any location", "New organization" etc.

nofaralfasi commented:
> I would say it's crucial to have it for the selected item on top. Let's ask in a PF4 react slack. It shouldn't be that difficult, but I don't know what component you are using. For example there is an icon within dropdown [here](patternfly.org/v4/components/menu-toggle/#with-iconimage-and-text). (May not apply to our problem)
> 
In the `ContextSelector` component that we are using,
the top selected text can be passed only as a string, therefor we can't pass an icon component.
We tried to pass Unicode Character “🌐” (U+1F310) though it didn't look well.

> Also, would it be possible to fix capitalization? We should use [sentence case](https://www.patternfly.org/v4/ux-writing/capitalization#capitalization-across-red-hat-uis) - so "Any organization", "Any location", "New organization" etc.

No problem, I will add that :+1:

MariSvirik commented:
I asked in the PF4 community and this was their answer:

"Even though the typescript type is set to string you could actually pass jsx like
<ContextSelector toggleText={
 <span><TimesIcon /> Link</span>
} />
but you’d have to silence the typescript error until the type is fixed"

Does this make any sense? @nofaralfasi 
(I would suggest joining patternfly-react channel in patternfly slack)

Also bug was filed here https://github.com/patternfly/patternfly-react/issues/7457

MariaAga commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -59,8 +61,15 @@ const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
-      {anyTaxonomyText}
+      <Grid hasGutter>
+        <GridItem span={1}>
+          <i className="fa fa-globe" aria-hidden="true" />
```
Can you use `<GlobeIcon />` here? its in `@patternfly/react-icons`.
Full list of icons is here: https://patternfly-react.surge.sh/icons/

Ron-Lavi replied:
The icons search in https://patternfly-react.surge.shworth nothing lol
thanks for finding this!

Ron-Lavi commented:
> Does this make any sense? @nofaralfasi

Actually we tried to pass an Icon React component but then in the UI it shows `[object object] Any organization`

MariaAga commented:
@Ron-Lavi @nofaralfasi you can do something like this:
``` js
 toggleText={
        currentTaxonomy || (
          <div>
            <GlobeIcon />
            {anyTaxonomyText}
          </div>
        )
      }
```
![Screenshot from 2022-05-25 12-27-29](https://user-images.githubusercontent.com/30431079/170241761-afc13f00-cf66-4fad-a5c6-1fa2f812d004.png)

nofaralfasi commented:
I added the globe icon as @MariaAga suggested, and also fixed the capitalization as @MariSvirik suggested.
You can see the changes in the screenshot:
![image](https://user-images.githubusercontent.com/38184193/170261664-fbf879da-a13c-434b-a06b-be0c40e6ab58.png)

Thanks for the help, and let me know if anything else is needed!

Ron-Lavi commented:
> @Ron-Lavi @nofaralfasi you can do something like this:
> ``` js
>  toggleText={
>         currentTaxonomy || (
>           <div>
>             <GlobeIcon />
>             {anyTaxonomyText}
>           </div>
>         )
>       }
> ```
> 

Oh that's great, I think what we tried was something like:
```js
`${icon} Any organization` 
```
And the component didn't know how to use it..

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -81,7 +90,14 @@ const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
+        currentTaxonomy || (
+          <div>
+            <GlobeIcon style={{ marginRight: '5px' }} />
+            {anyTaxonomyText}
```
Since we use the same icon and text in two places, can we store this whole JSX structure in a const and use it in the two places?

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -59,8 +61,15 @@ const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
-      {anyTaxonomyText}
+      <Grid hasGutter>
+        <GridItem span={1}>
+          <GlobeIcon style={{ marginRight: '5px' }} />
```
Can we also move the icon 1 or 2 pixels down?

Ron-Lavi reviewed (commented):
Thanks @nofaralfasi, almost there!

MariSvirik commented:
@nofaralfasi could you pls allign the text and icon? There is probably some extra CSS love needed to do it @jeremylenz was able to fix this problem in the host detail page.
Also, do we want to use the same icons for both?

nofaralfasi commented:
I added the building icon as @MariSvirik suggested, and also moved the icon 2 pixels down as @Ron-Lavi suggested.
You can see the changes in the screenshots:

![image](https://user-images.githubusercontent.com/38184193/170705429-dd758a0a-8345-4cbf-b120-d4c10b0e3a2f.png)

![image](https://user-images.githubusercontent.com/38184193/170705348-94bf95bc-2af2-406e-a5f5-e2813d93574f.png)

Let me know if any other changes are required.
Thanks

MariaAga commented:
When the js tests will work will need to update the snapshots (here or in a new pr)

nofaralfasi committed:
Fixes #28810 - changes to Any organization & Any location taxonomy's names

- Added unique name validations (English only)
- Added icons
- Fixed capitalization

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -49,6 +51,12 @@ const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
   const selectedIcon = <CheckIcon size="sm" className="current-taxonomy-v" />;
+  const anyIcon =
+    taxonomyType === 'organization' ? (
+      <BuildingIcon style={{ marginRight: '5px', marginTop: '2px' }} />
+    ) : (
+      <GlobeIcon style={{ marginRight: '5px', marginTop: '2px' }} />
```
Would another pixel down make sense? `marginTop: '3px'`

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -81,7 +94,14 @@ const TaxonomyDropdown = ({ taxonomyType, currentTaxonomy, taxonomies }) => {
-      toggleText={currentTaxonomy || anyTaxonomyText}
+      toggleText={
+        currentTaxonomy || (
+          <div>
+            {anyIcon}
+            {anyTaxonomyText}
+          </div>
```
we can use `React.Fragment` instead of a `div` which currently adds additional DOM element:
```suggestion
          <>
            {anyIcon}
            {anyTaxonomyText}
          </>
```

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/__snapshots__/TaxonomyDropdown.test.js.snap:
```diff
@@ -144,7 +304,23 @@ exports[`TaxonomyDropdown rendering rendering 1`] = `
+      <GridItem
+        span={1}
+      />
+    </Grid>
```
@MariaAga thoughts about removing this snap file and the test that generated it?

Ron-Lavi reviewed (dismissed):
thanks, @nofaralfasi!
LGTM, waiting for the tests results

Ron-Lavi commented:
ok to test

Ron-Lavi commented:
ok to test

Ron-Lavi commented:
@nofaralfasi the tests in Jenkins for Foreman and Katello keep failing with this PR :/

ofedoren commented:
Do we still want to have this check https://github.com/theforeman/foreman/pull/9222/files#diff-8a44d21124d0afdc8dffc797562785cbf95b4d8c79ca992669c841fa364f69b3R7 ? I guess @adamruzicka's comment about different languages is still valid, isn't it?..

nofaralfasi commented:
> @nofaralfasi the tests in Jenkins for Foreman and Katello keep failing with this PR :/

@Ron-Lavi You are right, thanks. I'm working on fixing that.

nofaralfasi commented:
> Do we still want to have this check https://github.com/theforeman/foreman/pull/9222/files#diff-8a44d21124d0afdc8dffc797562785cbf95b4d8c79ca992669c841fa364f69b3R7 ? I guess @adamruzicka's comment about different languages is still valid, isn't it?..

We decided to keep that validation anyway... So the validation will only work for English.
That is why we also added the icons, so in other languages `Any organization` and `Any location` will be unique.

Ron-Lavi reviewed (approved):
Thanks @nofaralfasi !
